pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,329 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 		readers = newChunkReaders(chunkReaders)
 	}
 
-	seriesSet, resHints, err := s.streamingSeriesSetForBlocks(ctx, req, blocks, indexReaders, readers, shardSelector, matchers, chunksLimiter, seriesLimiter, stats)
+	// If we are streaming the series labels and chunks separately, we don't need to fetch the postings
+	// twice. So we use these slices to re-use them.
+	// Each reusePostings[i] and reusePendingMatchers[i] corresponds to a single block.
+	var (
+		reusePostings        [][]storage.SeriesRef
+		reusePendingMatchers [][]*labels.Matcher
+		iterationBegin       time.Time
+	)
+	if req.StreamingChunksBatchSize > 0 {
+		// The streaming feature is enabled where we stream the series labels first, followed
+		// by the chunks later. Send only the labels here.
+		req.SkipChunks = true
+
+		reusePostings = make([][]storage.SeriesRef, len(blocks))
+		reusePendingMatchers = make([][]*labels.Matcher, len(blocks))
+		chunksLimiter := s.chunksLimiterFactory(s.metrics.queriesDropped.WithLabelValues("chunks"))
+		seriesLimiter := s.seriesLimiterFactory(s.metrics.queriesDropped.WithLabelValues("series"))
```

Comment:
Also maybe add a comment explaining that we create separate limiters so that fetching the series twice doesn't count twice to the limit.